### ISSUE TITLE
Support for Proxy Auth via. HTTP header

### ISSUE THREAD

**Issue by eyJhb** [wger-project/wger#1847]
## Use case
This would allow to put e.g. Authelia in front of Wger, and then use the header that Authelia passes to wger to create/signin the user.
This could function as a simple way, to  support the uses cases such as #1797 (not 100%, but it is quite simple to make).

See https://github.com/miniflux/v2/issues/534 and https://grafana.com/docs/grafana/latest/auth/auth-proxy/ .

## Proposal
- Add config option `PROXY_AUTH_HEADER=Remote-User`
- Add config option `PROXY_AUTH_CREATE_USER=True`
- `ALLOW_REGISTRATION=False` should still allow creating users using `PROXY_AUTH_`
- Mobile app should allow using API key instead of username/password for signin
  - This means user would first authenticate themselves using the website (go through e.g. Authelia -> Wger), copy API key, and then use that from the app.

I might take a jab at this if I have the time.

**Comment by rolandgeider** [comment]
interesting, this would mean taking the auth header and just going along with it. My feeling is that this should be rather easy to do (or there is already some package for this)

**Comment by eyJhb** [comment]
I remember a issue with this. So it would require the app to be changed as well, to allow using API key for logging in, instead of using username/password, as that would no longer work (they have no password).

My current nginx setup allows anyone to access `/static`, `/media` and `/api`, and that should still work with such a setup as this.

**Comment by rolandgeider** [comment]
(have started adding the setting and the docs to the other repos)

**Comment by rolandgeider** [comment]
grafana has an optional IP whitelist, might also be a good idea

**Comment by eyJhb** [comment]
> grafana has an optional IP whitelist, might also be a good idea

Might be a good idea to support in the future, but isn't necessary for this, is it? :)

**Comment by rolandgeider** [comment]
I just added it now, wasn't much work, seems more secure that way 😄 

If you can, can you checkout the branch and check that it works it with a real auth service? I've just done some very rudimentary tests

**Comment by eyJhb** [comment]
Is the intention that you're forced to set `AUTH_PROXY_TRUSTED_IPS`? Because it's somewhat dubious what it actually protects from. I think many will be running it through nginx, where `REMOTE_ADDR` should always be `127.0.0.1`, but that would STILL allow a malicious user to set the header, ie. no protection.

What it protects against, is if wger is bound to `0.0.0.0`, and someone connects directly to it, around the reverse proxy.

E.g. I specifically set this in places, where I want to ensure the client cannot manipulate the headers. This is e.g. for the API endpoint, which shouldn't use that form of authentication, on mobile devices.

```
proxy_set_header Remote-User "";
proxy_set_header Remote-Groups "";
proxy_set_header Remote-Email "";
proxy_set_header Remote-Name "";
```

I'll try to test it out when I find the time :)

**Comment by rolandgeider** [comment]
> What it protects against, is if wger is bound to 0.0.0.0, and someone connects directly to it, around the reverse proxy.

exactly. In the docker setup the application and the proxy are different services and you only want to accept the headers from the one you want (ok, here you couldn't access the application directly from the outside, but with other setups, you might). I feel that explicitly whitelisting the auth proxy makes it clearer and avoids potentially leaving this open because you didn't think about something.

> I'll try to test it out when I find the time :)

💪🏻

**Comment by tricepsbrachii** [comment]
I had a chance to try this with authentik proxy auth.

Good news: It works! I am adding my configuration details below in the hopes that it will help someone in the future.

<details>
<summary>Caddyfile</summary>

This is is constructed according to the standard authentik caddy template for a forward-auth proxy provider.
Note:
- `authentik_server` is the hostname of the authentik server
- `wger_server` is the hostname of the wger server
- I am not sure if the `header_up` statements are even required, I just copied that from the example caddyfile in the [wger-docker repo](https://github.com/wger-project/docker/blob/2f999f677b16a668daadc51508a9bbc70bb83c25/config/Caddyfile.example).
In this case, caddy is serving as the reverse proxy for wger. There is no additional proxy in-between (I am not using the nginx proxy from the default docker-compose recipe).
```
# Forward-auth proxy template for wger
# Using verbatim template from authentik for caddy
wger.example.com {
  import cert
    # Enclosing in `route` forces execution order
    route {
        # Forward outpost path to actual outpost
        reverse_proxy /outpost.goauthentik.io/* authentik_server:9000

        # Forward authentication to outpost
        forward_auth authentik_server:9000 {
            uri /outpost.goauthentik.io/auth/caddy

            # Header capitalization is case-sensitive
            copy_headers X-Authentik-Username X-Authentik-Groups X-Authentik-Entitlements X-Authentik-Email X-Authentik-Name X-Authentik-Uid X-Authentik-Jwt X-Authentik-Meta-Jwks X-Authentik-Meta-Outpost X-Authentik-Meta-Provider X-Authentik-Meta-App X-Authentik-Meta-Version

            # optional, in this config trust all private ranges, should probably be set to the outposts IP
            #trusted_proxies private_ranges
        }

        # site configuration
    handle /static/* {
        root * /wger
        file_server
    }

    handle /media/* {
        root * /wger
        file_server
   }
        reverse_proxy wger_server:8000 {
        header_up Host {host}
        header_up X-Real-IP {remote_host}
        header_up X-Forwarded-For {http.X-Forwarded-For} {remote_host}
        header_up X-Forwarded-Proto {scheme}
    }

  }
}
```
</details>

<details>
<summary>`prod.env` excerpt</summary>

```
#
# Auth Proxy Authentication
#
# Please read the documentation before enabling this feature:
# https://wger.readthedocs.io/en/latest/administration/auth_proxy.html
AUTH_PROXY_HEADER=HTTP_X_AUTHENTIK_NAME
AUTH_PROXY_TRUSTED_IPS=[insert IP of caddy host here]
AUTH_PROXY_CREATE_UNKNOWN_USER=True
AUTH_PROXY_USER_EMAIL_HEADER=HTTP_X_AUTHENTIK_EMAIL
AUTH_PROXY_USER_NAME_HEADER=
```
</details>

A few observations:
- If the user is not registered in wger already, I haven't been able to figure out how to get the unknown user creation working. Using the configuration above, I get the following error in the logs:
<details>
<summary>log exerpt</summary>

```
Jun 02 12:00:00 podman wger_server[12973]: level=ERROR ts=2025-06-02 12:00:00,344 module=backends path=/home/wger/src/wger/core/backends.py line=61 message=AuthProxy: Failed to create user 'testuser': null value in column "first_name" of relation "auth_user" violates not-null constraint
Jun 02 12:00:00 podman wger_server[12973]: DETAIL:  Failing row contains (13, !redacted_redacted_redacted, null, f, testuser, null, , user@0example.com, f, t, 2025-06-02 12:00:00.000+00).
Jun 02 12:00:00 podman wger_server[12973]: level=ERROR ts=2025-06-02 12:00:00,345 module=middleware path=/home/wger/src/wger/core/middleware.py line=129 message=AuthProxyMiddleware: Authentication failed for username 'testuser' from header 'HTTP_X_AUTHENTIK_NAME' (Trusted IP: 10.69.9.2).
```
</details>

- If the user is registered in wger already (with a username matching the username in authentik), then the proxy auth works as expected.
- It would be nice if `AUTH_PROXY_TRUSTED_IPS` accepted a subnet. Using podman, all of the wger containers in the pod are part of their own network. I trust all the containers in this subnet. The container corresponding to caddy is not guaranteed to have the same IP each time the pod is started (and I don't want to specify a static IP because the less hard-coded values the more maintainable the configuration is).

Otherwise, thank you for implementing this feature and thank you for the wger project!!! I am using it in my home cloud.

**Comment by eyJhb** [comment]
Great to hear that others are enjoying this feature as well! 

To fix the issue with not being able to create users with this, you need to configure the name option as well, e.g.

```
AUTH_PROXY_USER_NAME_HEADER=HTTP_X_AUTHENTIK_NAME
```

That should hopefully do the trick! :)

**Comment by tricepsbrachii** [comment]
That did the trick, now the user creation works! Thank you :)

**Comment by Gecko-with-a-hat** [comment]
Thanks to the helpful information on this page I was also able to get it working. 

I also second that it would be nice to set a subnet for the trusted IP. My entire setup will break if the docker ip of the reverse proxy changes. For now I will manually set the up for the reverse proxy in the compose file, but that's not a very good solution IMHO.

**Comment by rolandgeider** [comment]
> I also second that it would be nice to set a subnet for the trusted IP.

this sounds like a nice addition and should be easy enough to check with the `ipaddress` package. @Gecko-with-a-hat can you open a new issue?

**Comment by tricepsbrachii** [comment]
Just tested with the recently released wger-flutter release 1.8.3 and auth with the api key works as intended. Note to future visitors, make sure to bypass the api/* subfolder from your proxy auth so that the app can access the api endpoint. In caddy, it *might* look something like this:
```
wger.example.com {
  import cert
  # allow api path to bypass auth (requires api key)
  handle /api/* {
    reverse_proxy wger_server:8000
  }
    # Enclosing in `route` forces execution order
    route {
        # Forward outpost path to actual outpost
        reverse_proxy /outpost.goauthentik.io/* authentik_server:9000
  ...
```

**Comment by rolandgeider** [comment]
> Just tested with the recently released wger-flutter release 1.8.3 and auth with the api key works as intended. Note to future visitors, make sure to bypass the api/* subfolder from your proxy auth so that the app can access the api endpoint.

just added this to the docs :)
